### PR TITLE
[special] Add `cupyx.scipy.special.expn`

### PR DESCRIPTION
**Pull request by khushi-411**
#### TODO
- [x] add benchmarks

Benchmarks:
size | dtype | SciPy | CuPy
-----|-------|-------|-----
1000 | int8 | 0.119 ms | 0.203 ms
1000 | int32 | 0.111 ms | 0.201 ms
1000 | uint32 | 0.115 ms | 0.201 ms
1000 | float32 | 0.283 ms | 0.207 ms
1000 | float64 | 0.257 ms | 0.207 ms
100000 | int8 | 11.396 ms | 8.503 ms
100000 | int32 | 11.705 ms | 7.147 ms
100000 | uint32 | 11.495 ms | 6.202 ms
100000 | float32 | 17.490 ms | 6.235 ms
100000 | float64 | 17.482 ms | 6.227 ms
1000000 | int8 | 114.541 ms | 60.157 ms
1000000 | int32 | 110.311 ms | 58.221 ms
1000000 | uint32 | 109.507 ms | 58.139 ms
1000000 | float32 | 164.692 ms | 59.586 ms
1000000 | float64 | 165.868 ms | 62.705 ms

<details>
<summary>Script: Click Please</summary>

<p>

```python
import cupy
import cupyx

import scipy

from cupy import testing
from cupyx.profiler import benchmark


n_warmup = 1
n_repeat = 10
dtype_set = ('int8', 'int32', 'uint32', 'float32', 'float64')
n_set = (1000, 100000, 1000000)


def get_time(pref):
    cpu_time = pref.cpu_times.mean()
    gpu_time = pref.gpu_times.mean()
    return max(cpu_time, gpu_time) * 1000  # ms


def time_expn(a, b):
    cp_a = cupy.array(a)
    cp_b = cupy.array(b)
    ref = scipy.special.expn(a, b)
    ret = cupyx.scipy.special.expn(cp_a, cp_b)
    times = []
    perf = benchmark(scipy.special.expn, (a, b),
                     n_warmup=n_warmup, n_repeat=n_repeat)
    times.append(get_time(perf))
    perf = benchmark(cupyx.scipy.special.expn, (cp_a, cp_b),
                     n_warmup=n_warmup, n_repeat=n_repeat)
    times.append(get_time(perf))
    return times


print('size | dtype | SciPy | CuPy')
print('-----|-------|-------|-----')
for n in n_set:
    for dtype in dtype_set:
        a = scipy.random.randint(0, 2, size=n).astype(dtype)
        b = scipy.random.randint(0, 10, size=n).astype(dtype)
        times = time_expn(a, b)
        print('{} | {} | {:.3f} ms | {:.3f} ms'.format(n, dtype, times[0], times[1]))
```

</p>
</details>

### CODE REVIEW

File: cupyx/scipy/special/_expn.py [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+        return 0;
+    }
+
+    /* Do the k = 1 term outside the loop since A[1] = 1 */
+    fac *= multiplier;
+    res += fac;
+
+    for (k = 2; k < nA; k++) {
+        fac *= multiplier;
+        term = fac*polevl<Adegs[k]>(lambda, A[k]);
```
Comment (takagi) [inline cupyx/scipy/special/_expn.py:61]: Looks like that `polevl` in _digamma.py is templated by `int N` for optimization, giving the compiler a chance to unroll a loop in it, and you can't use `Adges[k]` for a template argument.

File: cupyx/scipy/special/_expn.py [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+	    qkm1 = qk;
+	    if (fabs(pk) > big) {
+	        pkm2 /= big;
+	        pkm1 /= big;
+	        qkm2 /= big;
+	        qkm1 /= big;
+	    }
+        } while (t > MACHEP);
+
+        ans *= exp(-x);
```
Comment (takagi) [inline cupyx/scipy/special/_expn.py:158]: Seems that you have to return `ans` without going through this block.

File: cupyx/scipy/special/_expn.py [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+__constant__ double A9[] = {362880.000000000000, -3733920.00000000000, 11026296.0000000000, -12440064.0000000000, 5765500.00000000000, -1062500.00000000000, 67260.0000000000000, -1004.00000000000000, 1.00000000000000000};
+__constant__ double A10[] = {-3628800.00000000000, 44339040.0000000000, -162186912.000000000, 238904904.000000000, -155357384.000000000, 44765000.0000000000, -5326160.00000000000, 218848.000000000000, -2026.00000000000000, 1.00000000000000000};
+__constant__ double A11[] = {39916800.0000000000, -568356480.000000000, 2507481216.00000000, -4642163952.00000000, 4002695088.00000000, -1648384304.00000000, 314369720.000000000, -25243904.0000000000, 695038.000000000000, -4072.00000000000000, 1.00000000000000000};
+__constant__ double A12[] = {-479001600.000000000, 7827719040.00000000, -40788301824.0000000, 92199790224.0000000, -101180433024.000000, 56041398784.0000000, -15548960784.0000000, 2051482776.00000000, -114876376.000000000, 2170626.00000000000, -8166.00000000000000, 1.00000000000000000};
+__constant__ double *A[] = {A0, A1, A2, A3, A4, A5, A6, A7, A8, A9, A10, A11, A12};
+__constant__ double Adegs[] = {0, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11};
```
Comment (takagi) [inline cupyx/scipy/special/_expn.py:37]:
```suggestion
__constant__ int Adegs[] = {0, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11};
```

File: cupyx/scipy/special/_expn.py (hunk 2 above)
Comment (khushi-411) [reply cupyx/scipy/special/_expn.py:293]: Ahh, yes! You are right. Have added it :)

File: cupyx/scipy/special/_expn.py [hunk 4]
```diff
@@ -0,0 +1,359 @@
...
+# Direct inquiries to 30 Frost Street, Cambridge, MA 02140
+
+from cupy import _core
+
+from cupyx.scipy.special._gamma import gamma_definition
+
+
+polevl_definition = '''
+
+static __device__ double polevl(double x, const double coef[], int N)
```
Comment (takagi) [inline cupyx/scipy/special/_expn.py:16]: Removing `const` qualifier here is fairly acceptable as this function is short and clear enough.

```suggestion
static __device__ double polevl(double x, double coef[], int N)
```

Comment (takagi) [inline cupyx/scipy/special/_expn.py:16]: Having `polevl` definition for `expn` here is okay!

Comment (takagi) [inline cupyx/scipy/special/_expn.py:16]: We don't need `static` storage-class specifier for CUDA device codes. I guess removing this resolves the ptxas fatal you mentioned.

```suggestion
__device__ double polevl(double x, const double coef[], int N)
```

Comment (khushi-411) [reply cupyx/scipy/special/_expn.py:16]: Great! Thanks for confirming :)

Comment (takagi) [inline cupyx/scipy/special/_expn.py:16]: Maybe removing `static` storage class resolves the ptxas fatal you mentioned.

```suggestion
__device__ double polevl(double x, const double coef[], int N)
```

Comment (khushi-411) [reply cupyx/scipy/special/_expn.py]: Actually yes! You are right! Thanks!

Comment (khushi-411) [reply cupyx/scipy/special/_expn.py]: Updated :)

**Comment by khushi-411** [comment]
I think this failure is unrelatable to the PR:
```python


Last 10 lines from the test output: |  
-- | --
00:49:43.379276 STDOUT 3240] | FAILED cupy_tests\core_tests\test_ndarray_adv_indexing.py::TestArrayAdvancedIndexingSetitemScalarValueDeprecated::test_adv_setitem[_param_8_{indexes=[[1, 0], 2], shape=(2, 3, 4), value=1}] |  
00:49:43.379775 STDOUT 3240] | FAILED cupy_tests\core_tests\test_ndarray_adv_indexing.py::TestArrayAdvancedIndexingSetitemScalarValueDeprecated::test_adv_setitem[_param_9_{indexes=[[1], slice(1, 2, None)], shape=(2, 3, 4), value=1}] |  
00:49:43.379775 STDOUT 3240] | FAILED cupy_tests\core_tests\test_ndarray_adv_indexing.py::TestArrayAdvancedIndexingSetitemScalarValueDeprecated::test_adv_setitem[_param_10_{indexes=[[[1]], slice(1, 2, None)], shape=(2, 3, 4), value=1}] |  
00:49:43.380175 STDOUT 3240] | FAILED cupy_tests\core_tests\test_ndarray_copy_and_view.py::TestArrayCopyAndView::test_view_f_contiguous |  
00:49:43.380175 STDOUT 3240] | FAILED cupy_tests\core_tests\test_ndarray_copy_and_view.py::TestArrayCopyAndView::test_view_flags_smaller[F-(3, 5)] |  
00:49:43.380175 STDOUT 3240] | FAILED cupy_tests\core_tests\test_ndarray_copy_and_view.py::TestArrayCopyAndView::test_view_flags_larger[F-(6, 5)] |  
00:49:43.380175 STDOUT 3240] | FAILED cupy_tests\core_tests\test_ndarray_copy_and_view.py::TestArrayCopyAndView::test_view_flags_larger[F-(2, 3)] |  
00:49:43.380175 STDOUT 3240] | FAILED cupy_tests\creation_tests\test_basic.py::TestBasic::test_empty_zero_sized_array_strides |  
00:49:43.380175 STDOUT 3240] | FAILED cupy_tests\functional_tests\test_vectorize.py::TestVectorizeOps::test_vectorize_div |  
00:49:43.380175 STDOUT 3240] | = 28 failed, 96133 passed, 5852 skipped, 689 deselected, 151 xfailed, 2569 warnings in 2247.59s (0:37:27) =

Last 10 lines from the test output:	
[00:49:43.379276 STDOUT 3240]](https://ci.preferred.jp/cupy.win.cuda116/104629/#L4877)	FAILED cupy_tests\core_tests\test_ndarray_adv_indexing.py::TestArrayAdvancedIndexingSetitemScalarValueDeprecated::test_adv_setitem[_param_8_{indexes=[[1, 0], 2], shape=(2, 3, 4), value=1}]	
[00:49:43.379775 STDOUT 3240]](https://ci.preferred.jp/cupy.win.cuda116/104629/#L4879)	FAILED cupy_tests\core_tests\test_ndarray_adv_indexing.py::TestArrayAdvancedIndexingSetitemScalarValueDeprecated::test_adv_setitem[_param_9_{indexes=[[1], slice(1, 2, None)], shape=(2, 3, 4), value=1}]	
[00:49:43.379775 STDOUT 3240]](https://ci.preferred.jp/cupy.win.cuda116/104629/#L4881)	FAILED cupy_tests\core_tests\test_ndarray_adv_indexing.py::TestArrayAdvancedIndexingSetitemScalarValueDeprecated::test_adv_setitem[_param_10_{indexes=[[[1]], slice(1, 2, None)], shape=(2, 3, 4), value=1}]	
[00:49:43.380175 STDOUT 3240]](https://ci.preferred.jp/cupy.win.cuda116/104629/#L4883)	FAILED cupy_tests\core_tests\test_ndarray_copy_and_view.py::TestArrayCopyAndView::test_view_f_contiguous	
[00:49:43.380175 STDOUT 3240]](https://ci.preferred.jp/cupy.win.cuda116/104629/#L4884)	FAILED cupy_tests\core_tests\test_ndarray_copy_and_view.py::TestArrayCopyAndView::test_view_flags_smaller[F-(3, 5)]	
[00:49:43.380175 STDOUT 3240]](https://ci.preferred.jp/cupy.win.cuda116/104629/#L4885)	FAILED cupy_tests\core_tests\test_ndarray_copy_and_view.py::TestArrayCopyAndView::test_view_flags_larger[F-(6, 5)]	
[00:49:43.380175 STDOUT 3240]](https://ci.preferred.jp/cupy.win.cuda116/104629/#L4886)	FAILED cupy_tests\core_tests\test_ndarray_copy_and_view.py::TestArrayCopyAndView::test_view_flags_larger[F-(2, 3)]	
[00:49:43.380175 STDOUT 3240]](https://ci.preferred.jp/cupy.win.cuda116/104629/#L4887)	FAILED cupy_tests\creation_tests\test_basic.py::TestBasic::test_empty_zero_sized_array_strides	
[00:49:43.380175 STDOUT 3240]](https://ci.preferred.jp/cupy.win.cuda116/104629/#L4888)	FAILED cupy_tests\functional_tests\test_vectorize.py::TestVectorizeOps::test_vectorize_div	
[00:49:43.380175 STDOUT 3240]](https://ci.preferred.jp/cupy.win.cuda116/104629/#L4889)	= 28 failed, 96133 passed, 5852 skipped, 689 deselected, 151 xfailed, 2569 warnings in 2247.59s (0:37:27) =
```

**Comment by takagi** [comment]
Yes, its because of NumPy 1.23 that is recently released, and I'm working on fixing it.

**Review by takagi — COMMENTED** [review]
LGTM!

**Review by takagi — APPROVED** [review]
LGTM!

**Comment by khushi-411** [comment]
Thanks a lot for your suggestions and guidance! :)

**Comment by takagi** [comment]
Would you merge this as the failure of `cupy.win.cuda116` is not related. @kmaehashi

**Comment by takagi** [comment]
Thanks for your work, @khushi-411!

**Pull request merged**